sound/soc/sof/ipc4-topology.c

```diff
@@ -168,6 +168,75 @@ static void sof_ipc4_dbg_audio_format(struct device *dev,
 	}
 }
 
+static const struct sof_ipc4_audio_format *
+sof_ipc4_get_widget_pin_format(struct snd_sof_widget *swidget, bool pin_type,
+			       int pin_id)
+{
+	struct snd_soc_dapm_widget *widget = swidget->widget;
+	struct sof_ipc4_base_module_cfg_ext *base_ext = NULL;
+	const struct sof_ipc4_pin_format *pin_fmt;
+	struct sof_ipc4_control_data *control_data;
+	const struct snd_kcontrol_new *kc;
+	struct snd_sof_control *scontrol;
+	struct soc_bytes_ext *sbe;
+	int pin_start, pin_end;
+	int i;
+
+	if (swidget->init_blob_style == SOF_INIT_BLOB_BASE_WITH_MOD_EXT) {
+		struct sof_ipc4_base_module_cfg *base_config = swidget->private;
+
+		if (pin_type == SOF_PIN_TYPE_SOURCE && swidget->payload_with_output_fmt) {
+			/* Output format follows base module config */
+			const struct sof_ipc4_audio_format *out_fmt = (void *)base_config +
+				sizeof(*base_config);
+			return out_fmt;
+		}
+
+		return &base_config->audio_fmt;
+	}
+
+	/*
+	 * Return the pin format from sof_ipc4_base_module_cfg_ext structure if module
+	 * uses base module config + base module config extension init blob.
+	 */
+	for (i = 0; i < widget->num_kcontrols; i++) {
+		kc = &widget->kcontrol_news[i];
+
+		/* payload uses byte kcontrol */
+		if (widget->dobj.widget.kcontrol_type[i] != SND_SOC_TPLG_TYPE_BYTES)
+			continue;
+		sbe = (struct soc_bytes_ext *)kc->private_value;
+		scontrol = sbe->dobj.private;
+		if (!scontrol)
+			continue;
+		control_data = scontrol->ipc_control_data;
+		if (control_data->data->blob_type != SOF_IPC4_MOD_INIT_INSTANCE)
+			continue;
+
+		base_ext = (struct sof_ipc4_base_module_cfg_ext *)control_data->data->data;
+		break;
+	}
+
+	if (!base_ext)
+		return NULL;
+
+	pin_fmt = base_ext->sink_pin_fmt;
+	if (pin_type == SOF_PIN_TYPE_SINK) {
+		pin_start = 0;
+		pin_end = base_ext->num_sink_pin_fmts;
+	} else {
+		pin_start = base_ext->num_sink_pin_fmts;
+		pin_end = base_ext->num_sink_pin_fmts + base_ext->num_source_pin_fmts;
+	}
+
+	for (i = pin_start; i < pin_end; i++) {
+		if (pin_fmt[i].pin_index == pin_id)
```

Comment:
you've got to explain how the pins are ordered or numbered. This is super hard to review.